File: backend_py/services/ml_service.py
import aiohttp
from typing import Dict, Any, Optional
from loguru import logger
from pydantic import BaseModel

class MLPrediction(BaseModel):
    prediction: int  # 0 or 1
    confidence: float
    signal: str  # 'BUY' | 'SELL' | 'HOLD'
    timestamp: str

class MLService:
    def __init__(self, base_url: str, timeout: int = 10):
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.is_available = False
        logger.info(f'ML Service initialized: {base_url}')

    async def check_health(self) -> bool:
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(f'{self.base_url}/health', timeout=self.timeout) as response:
                    if response.status == 200:
                        data = await response.json()
                        self.is_available = data.get('status') == 'ok' and data.get('model_loaded', False)
                        if self.is_available:
                            logger.info('✅ ML Service is available and model is loaded')
                        else:
                            logger.warning('⚠️ ML Service is available but model is not loaded')
                        return self.is_available
                    else:
                        logger.warning(f'⚠️ ML Service health check failed with status {response.status}')
                        self.is_available = False
                        return False
        except Exception as error:
            logger.warning(f'⚠️ ML Service is not available: {error}')
            self.is_available = False
            return False

    async def get_prediction(self, market_data: Dict[str, Any]) -> Optional[MLPrediction]:
        if not self.is_available:
            return None

        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f'{self.base_url}/predict',
                    json=market_data,
                    timeout=self.timeout
                ) as response:
                    if response.status == 200:
                        data = await response.json()
                        return MLPrediction(**data)
                    else:
                        logger.error(f'ML prediction failed with status {response.status}')
                        return None
        except Exception as error:
            logger.error(f'Failed to get ML prediction: {error}')
            return None

    async def get_model_info(self) -> Optional[Dict[str, Any]]:
        if not self.is_available:
            return None

        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(f'{self.base_url}/model/info', timeout=self.timeout) as response:
                    if response.status == 200:
                        return await response.json()
                    else:
                        logger.error(f'Failed to get model info with status {response.status}')
                        return None
        except Exception as error:
            logger.error(f'Failed to get model info: {error}')
            return None

File: src/services/mlService.ts
import axios, { AxiosInstance } from 'axios';
import { logger } from '../utils/logger';

export interface MLPrediction {
  prediction: 0 | 1;
  confidence: number;
  signal: 'BUY' | 'SELL' | 'HOLD';
  timestamp: string;
}

export interface MLServiceConfig {
  baseURL: string;
  timeout?: number;
}

export class MLService {
  private client: AxiosInstance;
  private isAvailable: boolean = false;

  constructor(config: MLServiceConfig) {
    this.client = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout || 10000,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    logger.info(`ML Service initialized: ${config.baseURL}`);
  }

  /**
   * Проверка доступности ML сервиса
   */
  async checkHealth(): Promise<boolean> {
    try {
      const response = await this.client.get('/health');
      this.isAvailable = response.data.status === 'ok' && response.data.model_loaded;

      if (this.isAvailable) {
        logger.info('✅ ML Service is available and model is loaded');
      } else {
        logger.warn('⚠️ ML Service is available but model is not loaded');
      }

      return this.isAvailable;
    } catch (error) {
      logger.warn('⚠️ ML Service is not available:', (error as Error).message);
      this.isAvailable = false;
      return false;
    }
  }

  /**
   * Получить ML предсказание
   */
  async predict(features: number[], ohlcv?: any[]): Promise<MLPrediction> {
    try {
      const response = await this.client.post('/predict', {
        features,
        ohlcv: ohlcv || []
      });

      return {
        prediction: response.data.prediction,
        confidence: response.data.confidence,
        signal: response.data.signal,
        timestamp: response.data.timestamp
      };
    } catch (error) {
      logger.error('Failed to get ML prediction:', error);

      // Возвращаем нейтральное предсказание в случае ошибки
      return {
        prediction: 0,
        confidence: 0.5,
        signal: 'HOLD',
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Проверить, доступен ли ML сервис
   */
  isServiceAvailable(): boolean {
    return this.isAvailable;
  }

  /**
   * Подготовить features из OHLCV данных
   * (упрощенная версия, можно расширить)
   */
  prepareFeatures(ohlcv: any[]): number[] {
    if (ohlcv.length < 20) {
      return [];
    }

    const closes = ohlcv.map(c => c.close);
    const volumes = ohlcv.map(c => c.volume);

    // Простые features для ML
    const features: number[] = [];

    // 1. Последние 5 цен закрытия (нормализованные)
    const recentCloses = closes.slice(-5);
    const avgClose = recentCloses.reduce((a, b) => a + b, 0) / recentCloses.length;
    recentCloses.forEach(close => {
      features.push((close - avgClose) / avgClose);
    });

    // 2. Изменение цены за последние 5 периодов
    for (let i = 1; i <= 5; i++) {
      const change = (closes[closes.length - i] - closes[closes.length - i - 1]) / closes[closes.length - i - 1];
      features.push(change);
    }

    // 3. Средний объем
    const avgVolume = volumes.slice(-10).reduce((a, b) => a + b, 0) / 10;
    features.push(volumes[volumes.length - 1] / avgVolume - 1);

    // 4. Волатильность (стандартное отклонение последних 10 цен)
    const last10 = closes.slice(-10);
    const mean = last10.reduce((a, b) => a + b, 0) / last10.length;
    const variance = last10.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / last10.length;
    features.push(Math.sqrt(variance) / mean);

    return features;
  }
}
